src/hotspot/share/asm/assembler.hpp

```diff
@@ -298,15 +298,17 @@ class AbstractAssembler : public ResourceObj  {
   // ensure buf contains all code (call this before using/copying the code)
   void flush();
 
-  void emit_int8(   uint8_t x1)                                     { code_section()->emit_int8(x1); }
+  // These take full width arguments and cast down without checking. Many callers have sign extension converting up to
+  // full int width arguments, so this casts this away.
+  void emit_int8(   uint32_t x1)                                    { code_section()->emit_int8((uint8_t)x1); }
```

Comment:
It is unfortunate that emit_int8 has that name, rather than emit_uint8.
But that's probably water under the bridge; certainly so for this change.

It is also unfortunate that integral promotion and -Wconversion fight each
other pretty hard.  And implicit conversions make it easy for -Wconversion
failure cases to slip in while we don't have it enabled as part of normal
operations. Here's an alternative approach:

```
template<typename T, ENABLE_IF(std::is_same<T, uint8_t>::value)>
void emit_int8(T v) { ... }

template<typename T,
         ENABLE_IF(std::is_integral<T>::value),
         ENABLE_IF(!std::is_same<T, uint8_t>::value)>
void emit_int8(T v) { emit_int8(checked_cast<uint8_t>(v)); }
```

Maybe that's not so bad?

Reply:
Thanks for the suggested change.  The sign extension for integral promotion was a problem.  The second emit_int8 will still assert for the callers (there are still some) that pass a type that gets sign extended to int to emit_in8.  I can see how many are left once the first function is added.  Maybe it's possible to fix them all.  We'll see.

Reply:
src/hotspot/cpu/x86/assembler_x86.cpp|4421 col 12| error: no matching function for call to 'Assembler::emit_int8(Assembler::ComparisonPredicate&)'                                                                                                                                          
||  4421 |   emit_int8(vcc);
||       |   ~~~~~~~~~^~~~~
|| In file included from /scratch/cphillim/hg/21more-conversion/src/hotspot/cpu/x86/assembler_x86.cpp:26:

But I made ComparisonPredicate enum a uint8_t.  Why didn't it match it?

  // Comparison predicates for integral types & FP types when using SSE
  enum ComparisonPredicate : uint8_t {

There are a bunch of these that don't like the templates.

Reply:
Taking out the ENABLE_IF<is_integral> at least allows it to compile but there are lots of crashes because of sign extension.

Maybe we can make checked_cast<> tolerant of that somehow?

```template <typename T2, typename T1>
constexpr T2 checked_cast(T1 thing) {
  T2 result = static_cast<T2>(thing);
  assert(static_cast<T1>(result) == thing, "must be");
  return result;
}```

Reply:
How about something like this?
```  template<typename T, ENABLE_IF(std::is_same<T, uint8_t>::value)>
  void emit_int8(T v) { code_section()->emit_int8(v); }

  template<typename T, ENABLE_IF(!std::is_same<T, uint8_t>::value)>
  void emit_int8(T v) {
    if (v < 0) {
      int8_t tmp = static_cast<int8_t>(v);
      assert(static_cast<T>(v) == tmp, "signed cast preserves value");
      emit_int8(static_cast<uint8_t>(v));
    } else {
      emit_int8(checked_cast<uint8_t>(v)); 
    }
  }
```